### PR TITLE
Display NFTs

### PR DESCRIPTION
**Pull request by TusharPardhe**
This MR adds 
- NFT displaying functionality
- View Button to redirect to NFT details
- Lazy loading of images
- Infinite scrolling for NFT listings

### CODE REVIEW

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx [hunk 1]
```diff
@@ -75,15 +87,15 @@ const LedgerProvider: FC = ({ children }) => {
...
+          account: 'rDGh681kc6V1GKkQB378XhiM1tzkYrnFwQ', // Change before merging to master
           limit: payload?.limit,
           marker: payload?.marker,
           ledger_index: 'validated'
         });
         if (!prepared.result?.account_nfts) {
           throw new Error("Couldn't get the NFTs");
         } else {
-          return prepared.result.account_nfts;
+          return { account_nfts: prepared.result.account_nfts, marker: prepared.result.marker };
```
Comment (TusharPardhe) [inline packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:98]: This issue has been addressed in an earlier MR

File: .vscode/settings.json [hunk 2]
```diff
@@ -15,6 +15,6 @@
   },
   "cSpell.words": ["gemwallet"],
   "[typescriptreact]": {
-    "editor.defaultFormatter": "esbenp.prettier-vscode"
+    "editor.defaultFormatter": "vscode.typescript-language-features"
```
Comment (FlorianBouron) [inline .vscode/settings.json:18]: What is the reason behind changing to this formater?

File: packages/extension/package.json [hunk 3]
```diff
@@ -40,7 +42,8 @@
     "@types/node": "^12.0.0",
     "@types/react": "^17.0.0",
     "@types/react-dom": "^17.0.0",
-    "cypress": "^12.9.0",
+    "@types/react-lazy-load-image-component": "^1.5.2",
+    "cypress": "^12.5.1",
```
Comment (FlorianBouron) [inline packages/extension/package.json]: You have downgraded the cypress version, could you please revert this change?

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx (hunk 1 above)
Comment (FlorianBouron) [reply packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:98]: I will have a look at it, I think the previous change is changing the API, this would create a breaking change, we cannot update it as is without an API release, let me check this in your other MR.

**Comment by FlorianBouron** [comment]
Could you please rebase your branch instead of `merging` from master? This is creating a cleaner git history.
I have cherry-picked your commits to clean up the git-history, please pull the branch again before making changes, the git history has been changed.

**Comment by FlorianBouron** [comment]
The cypress tests are failing on the pipeline, could you please make them pass?
Also you will need to check that the NFT viewer feature is fully passing with cypress to make sure we do not have any regressions in the future :)

File: packages/extension/package.json (hunk 3 above)
Comment (FlorianBouron) [reply packages/extension/package.json]: I have fixed it during the rebase interfactive

File: packages/constants/src/payload/payload.types.ts [hunk 4]
```diff
@@ -52,6 +52,10 @@ export interface SignMessageRequestPayload {
   message: string;
 }
 
+export interface NftImageRequestPayload {
```
Comment (FlorianBouron) [inline packages/constants/src/payload/payload.types.ts:55]: This is not shared with the API somehow, I don't think it should live here,  you should keep it within the component.

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx [hunk 5]
```diff
@@ -250,13 +262,41 @@ const LedgerProvider: FC = ({ children }) => {
...
+
+      if (url.length === 0) {
+        throw new Error('URI is empty');
+      }
+
+      url = url.replace('ipfs://', 'https://ipfs.io/ipfs/');
+      const nftData = await fetch(url)
+        .then((res) => res.json())
+        .catch(() => ({
+          name: '-',
```
Comment (FlorianBouron) [inline packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:278]: Instead of putting a dash, maybe we could write something generic like: `No name`, or even better return null or undefined in order to don't display it in the UI?

File: packages/extension/src/components/pages/Nfts/Nfts.tsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  account_nfts: [],
+  marker: null
+};
+
+export const Nfts: FC = () => {
+  const { getNFTs } = useContext(LedgerContext);
+
+  const [nfts, setNfts] = useState<AccountNFTokenResponse>(initalState);
+
+  const fetchNfts = async () => {
```
Comment (FlorianBouron) [inline packages/extension/src/components/pages/Nfts/Nfts.tsx:19]: This can be defined within your useEffect.
If you let it here, the function is gonna be re-defined at each render of the component.
See an example here: https://github.com/GemWallet/gemwallet-extension/blob/master/packages/extension/src/components/pages/SendPayment/PreparePayment/PreparePayment.tsx#L55

File: packages/extension/src/components/pages/Nfts/Nfts.tsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      const payload = {
+        limit: 20,
+        marker: nfts.marker
+      };
+
+      if (!nfts.marker) {
+        delete payload.marker;
+      }
+
+      const response = await getNFTs(payload);
```
Comment (FlorianBouron) [inline packages/extension/src/components/pages/Nfts/Nfts.tsx:30]: I would simplify this code that way:

```suggestion
      const response = await getNFTs({
            limit: 20,
            marker: nfts.marker ?? undefined
      );
```

File: packages/extension/src/components/organisms/NftListing/NftListing.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+import { InformationMessage } from '../../molecules';
+import { NftCard } from '../../molecules/NftCard';
+
+export interface NftListingProps extends AccountNFTokenResponse {
+  onLoadMoreClick: () => void;
+}
+
+export const NftListing: FC<NftListingProps> = ({ account_nfts, marker, onLoadMoreClick }) => {
+  if (account_nfts.length === 0) {
```
Comment (FlorianBouron) [inline packages/extension/src/components/organisms/NftListing/NftListing.tsx:16]: You need a loading state, when I am going to the NFTListing page, I can see the InformationMessage and then a list of NFTs.

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx [hunk 9]
```diff
@@ -75,15 +87,15 @@ const LedgerProvider: FC = ({ children }) => {
         // Prepare the transaction
         const prepared = await client.request({
           command: 'account_nfts',
-          account: wallet.publicAddress,
+          account: 'rDGh681kc6V1GKkQB378XhiM1tzkYrnFwQ', // Change before merging to master
```
Comment (FlorianBouron) [inline packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:90]: I think you forgot to revert this change :)

File: .vscode/settings.json (hunk 2 above)
Comment (TusharPardhe) [reply .vscode/settings.json:18]: Reverted

File: packages/constants/src/payload/payload.types.ts (hunk 4 above)
Comment (TusharPardhe) [reply packages/constants/src/payload/payload.types.ts:55]: Moved

File: packages/extension/src/components/pages/Nfts/Nfts.tsx (hunk 6 above)
Comment (TusharPardhe) [reply packages/extension/src/components/pages/Nfts/Nfts.tsx:19]: Here we need to pass this function to the next component as well so defined it outside the useEffect

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx (hunk 9 above)
Comment (TusharPardhe) [reply packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:90]: Just keeping here for testing until everything else is ready, will remove it after that :)

File: packages/extension/src/components/pages/Nfts/Nfts.tsx (hunk 7 above)
Comment (TusharPardhe) [reply packages/extension/src/components/pages/Nfts/Nfts.tsx:30]: added

File: packages/extension/src/components/organisms/NftListing/NftListing.tsx (hunk 8 above)
Comment (TusharPardhe) [reply packages/extension/src/components/organisms/NftListing/NftListing.tsx]: added

**Comment by ThibautBremand** [comment]
@FlorianBouron to be tested with a wallet with a lot of NFTs

File: packages/extension/src/utils/NFTViewer.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+  return {
+    ...NFTData,
+    NFTokenID,
+    image
+  };
+};
+
+const replaceIPFS = (inputStr: string): string => {
+  if (!inputStr.startsWith('ipfs://' || 'https://ipfs.io/ipfs/')) {
```
Comment (ThibautBremand) [inline packages/extension/src/utils/NFTViewer.ts:30]: Need to also check that is does not start with http://

Comment (ThibautBremand) [reply packages/extension/src/utils/NFTViewer.ts:30]: Also need to handle stringified urls such as https:\/\/kegsrp.com\/nft\/img\/2333.png

File: packages/extension/src/components/pages/NFTs/NFTs.tsx [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        marker: NFTs.marker ?? undefined
+      };
+
+      setNFTs({ ...NFTs, loading: true });
+
+      const response = await getNFTs(payload);
+
+      setNFTs({
+        marker: response.marker,
+        account_nfts: NFTs.account_nfts.concat(response.account_nfts),
```
Comment (florent-uzio) [inline packages/extension/src/components/pages/NFTs/NFTs.tsx:39]: Why do we need to concat? Isn't the initial state an empty array?

Would this be simpler?

```jsx
const { marker, account_nfts } = await getNFTs(payload);

setNFTs({
        marker,
        account_nfts,
        loading: false
});
```

**Review by florent-uzio — CHANGES_REQUESTED** [review]
Nice job!

I am requesting some changes for some code optimization. Let me know if you have any questions.

**Comment by FlorianBouron** [comment]
@all-contributors add @florent-uzio for idea

File: packages/extension/src/utils/format.ts [hunk 12]
```diff
@@ -93,3 +93,28 @@ export const formatFlagsToNumber = (tx: Transaction) => {
 export const formatTransferFee = (fee: number) => {
   return fee / 1000;
 };
+
+// If the given string is longer than the given length, and does not contain a space within the length,
+// then it will be truncated and an ellipsis will be added to the end.
+// We do this, because if there is a space within the range, the string will be correctly displayed in the UI.
+export const formatStringToBeDisplayed = (str: string, length: number) => {
```
Comment (FlorianBouron) [inline packages/extension/src/utils/format.ts:100]: This is in the util folder, the function doesn't have a generic explicit name, I would rename it, not sure to understand what it does without the help of the comments and reading all of it.

Comment (ThibautBremand) [reply packages/extension/src/utils/format.ts:100]: Here is what the function does

When we display some strings to the UI, we want to make sure that they are not too long and will not break the UI.
To do so, we truncate the string only if:
 - There is one word with a length bigger than the given length
 otherwise, we return the string as is.  
 
We do not need to truncate the string if there is no word with a length bigger than the given length because in this
 case, the string will be correctly displayed in multiple lines, and will not break the UI.

Comment (ThibautBremand) [reply packages/extension/src/utils/format.ts:103]: Here is a case where the string (NFT description) is longer than the length of the UI, but it is not truncated because it does not contain long words, so it does not break the UI. If it contained a long word, such as a NFT Token ID, it would break the UI.   

Screenshot 2023-08-06 at 22 10 05

File: packages/extension/src/components/pages/NFTs/NFTs.tsx (hunk 11 above)
Comment (ThibautBremand) [reply packages/extension/src/components/pages/NFTs/NFTs.tsx]: It's because everytime we fetch more NFTs using the marker, we concatenate them with the previous NFTs that have already been fetched

```
<NFTListing
        {...{
          ...NFTs,
          onLoadMoreClick: fetchNFTs
        }}
      />
```

```
  const fetchNFTs = async () => {
    try {
      const payload = {
        limit: MAX_FETCHED_NFTS,
        marker: NFTs.marker ?? undefined
      };

      setNFTs({ ...NFTs, loading: true });

      const response = await getNFTs(payload);

      setNFTs({
        marker: response.marker, // --> set the marker here
        account_nfts: NFTs.account_nfts.concat(response.account_nfts),
        loading: false
      });
    } catch (error) {
      setNFTs(initalState);
    }
```

File: packages/extension/src/components/pages/NFTs/NFTs.tsx [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  loading: boolean;
+}
+
+export const NFTs: FC = () => {
+  const { getNFTs } = useContext(LedgerContext);
+
+  const [NFTs, setNFTs] = useState<NFTsProps>(initalState);
+
+  const fetchNFTs = async () => {
+    try {
```
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/extension/src/components/pages/NFTs/NFTs.tsx]: Right now all the methods in the LedgerContext throw exceptions on error, so I think we can keep it this way until we decide to change this to the complete LedgerContext

File: packages/extension/cypress/e2e/NFT.cy.ts [hunk 14]
```diff
@@ -7,7 +7,7 @@ import { formatFlags } from '../../src/utils';
 describe('Mint', () => {
   // deepcode ignore NoHardcodedPasswords: password used for testing purposes
   const PASSWORD = 'SECRET_PASSWORD';
-  const MINT_NFT_URL = `http://localhost:3000?mint-nft?URI=4d696e746564207468726f7567682047656d57616c6c657421&flags=%7B%22tfOnlyXRP%22%3Afalse%2C%22tfTransferable%22%3Atrue%7D&fee=199&transferFee=3000&NFTokenTaxon=0&memos=%5B%7B%22memo%22%3A%7B%22memoType%22%3A%224465736372697074696f6e%22%2C%22memoData%22%3A%2254657374206d656d6f%22%7D%7D%5D&id=210324818&requestMessage=undefined&transaction=mintNFT`;
+  const MINT_NFT_URL = `http://localhost:3000?mint-nft&URI=4d696e746564207468726f7567682047656d57616c6c657421&flags=%7B%22tfOnlyXRP%22%3Afalse%2C%22tfTransferable%22%3Atrue%7D&fee=199&transferFee=3000&NFTokenTaxon=0&memos=%5B%7B%22memo%22%3A%7B%22memoType%22%3A%224465736372697074696f6e%22%2C%22memoData%22%3A%2254657374206d656d6f%22%7D%7D%5D&id=210324818&requestMessage=undefined&transaction=mintNFT`;
```
Comment (FlorianBouron) [inline packages/extension/cypress/e2e/NFT.cy.ts:10]: It seems odd to me, shouldn't it be:
```suggestion
  const MINT_NFT_URL = `http://localhost:3000/mint-nft?URI=4d696e746564207468726f7567682047656d57616c6c657421&flags=%7B%22tfOnlyXRP%22%3Afalse%2C%22tfTransferable%22%3Atrue%7D&fee=199&transferFee=3000&NFTokenTaxon=0&memos=%5B%7B%22memo%22%3A%7B%22memoType%22%3A%224465736372697074696f6e%22%2C%22memoData%22%3A%2254657374206d656d6f%22%7D%7D%5D&id=210324818&requestMessage=undefined&transaction=mintNFT`;
```

File: packages/extension/cypress/e2e/wallet_management.cy.ts [hunk 15]
```diff
@@ -481,7 +481,7 @@ describe('Edit wallet', () => {
 
     // Lock the extension
     cy.get('button[aria-label="close"]').click();
-    cy.contains('button', 'Wallets').parent().children().eq(2).click();
+    cy.contains('button', 'Wallets').parent().children().eq(3).click();
```
Comment (FlorianBouron) [inline packages/extension/cypress/e2e/wallet_management.cy.ts:484]: Why has this changed? Some comments for bellow.

File: packages/extension/src/App.tsx [hunk 16]
```diff
@@ -510,6 +512,14 @@ const App: FC = () => {
             </PrivateRoute>
           }
         />
+        <Route
```
[the remark this answers is not in the document]
Comment (FlorianBouron) [reply packages/extension/src/App.tsx:525]: Thank you @florent-uzio

File: packages/extension/src/utils/NFTImageResolver.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
+import { convertHexToString } from 'xrpl';
+
+import { AccountNFToken, NFTData } from '@gemwallet/constants';
+
+import { parseJSON } from './NFTViewer';
+
+export const resolveNFTImage = async (NFT: AccountNFToken): Promise<NFTData> => {
+  const { NFTokenID, URI } = NFT;
+  let URL = URI ? await convertHexToString(String(URI)) : '';
```
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/extension/src/utils/NFTImageResolver.ts]: To me, only `let URL = URI ? await convertHexToString(String(URI)) : '';` makes the resolver working locally

File: packages/extension/src/components/pages/NFTViewer/NFTViewer.tsx [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+interface NFTsProps extends AccountNFTokenResponse {
+  loading: boolean;
+}
+
+export const NFTViewer: FC = () => {
+  const { getNFTs } = useLedger();
+
+  const [NFTs, setNFTs] = useState<NFTsProps>(initalState);
+
+  const fetchNFTs = async () => {
```
Comment (FlorianBouron) [inline packages/extension/src/components/pages/NFTViewer/NFTViewer.tsx:26]: Instead of redefining this at every-lunch, we could either wrap it in a useCallback, or you could use it only in the useEffect.
You can find an example of async use here: packages/extension/src/contexts/NetworkContext/NetworkContext.tsx:47

File: packages/extension/cypress/e2e/wallet_management.cy.ts (hunk 15 above)
Comment (ThibautBremand) [reply packages/extension/cypress/e2e/wallet_management.cy.ts:484]: "NFTs" button has been added to the menu at position 2

File: packages/extension/src/components/pages/NFTViewer/NFTViewer.tsx [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        loading: false
+      });
+    } catch (error) {
+      setNFTs(initalState);
+    }
+  };
+
+  useEffect(() => {
+    fetchNFTs();
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- we only want to fetch once
```
Comment (FlorianBouron) [inline packages/extension/src/components/pages/NFTViewer/NFTViewer.tsx:49]: If there are dependencies like: `getNFTs`
You can safely pass them here :)

File: packages/extension/src/components/pages/NFTViewer/NFTViewer.tsx [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+
+      const response = await getNFTs(payload);
+
+      setNFTs({
+        marker: response.marker,
+        account_nfts: NFTs.account_nfts.concat(response.account_nfts),
+        loading: false
+      });
+    } catch (error) {
+      setNFTs(initalState);
```
Comment (FlorianBouron) [inline packages/extension/src/components/pages/NFTViewer/NFTViewer.tsx:43]: I think we should send an error to Sentry if something is wrong here, it will help us to troubleshot.

File: packages/extension/src/components/pages/NFTViewer/NFTViewer.tsx (hunk 19 above)
Comment (ThibautBremand) [reply packages/extension/src/components/pages/NFTViewer/NFTViewer.tsx:51]: The dependency would be fetchNFTs only here

File: packages/extension/src/components/organisms/NFTDetails/NFTDetails.tsx [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+              src={NFTData?.image}
+              width={250}
+            />
+          </div>
+        ) : (
+          <div style={{ display: 'flex', justifyContent: 'center', alignItems: 'center' }}>
+            <GemWallet />
+          </div>
+        )}
+        <ListItem style={{ padding: '8px 24px' }}>
```
Comment (florent-uzio) [inline packages/extension/src/components/organisms/NFTDetails/NFTDetails.tsx:63]: Those ListItem should be moved to a reusable component as we repeat the logic below.

**Review by florent-uzio — CHANGES_REQUESTED** [review]
Thanks for all the changes, nice work.

I am requesting changes to refactor some code by creating reusable components.
Let me know if you have any questions.

File: packages/extension/src/components/molecules/NFTCard/NFTCard.tsx [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+const Transition = forwardRef(function Transition(
+  props: TransitionProps & {
+    children: React.ReactElement;
+  },
+  ref: React.Ref<unknown>
+) {
+  return <Slide direction="up" ref={ref} {...props} />;
+});
+
+export const NFTCard: FC<NFTCardProps> = ({ NFT }) => {
```
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/extension/src/components/molecules/NFTCard/NFTCard.tsx:30]: We did not take this standard into account, thank you for the insight 👍 
Maybe you have a NFT Token ID that follows this standard? Anyway we need to tackle this

[the remark this answers is not in the document]
Comment (florent-uzio) [reply packages/extension/src/components/molecules/NFTCard/NFTCard.tsx:30]: Mmmh no sorry I don't have an NFT with that standard.

File: packages/extension/src/utils/format.ts (hunk 12 above)
Comment (ThibautBremand) [reply packages/extension/src/utils/format.ts]: I have found a way to do without this function, with a new component called <TruncatedText> 👍

File: packages/extension/src/components/molecules/NFTCard/NFTCard.tsx [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+import { GemWallet } from '../../atoms';
+import { NFTDetails } from '../../organisms';
+
+export interface NFTCardProps {
+  NFT: AccountNFToken;
+}
+
+const Transition = forwardRef(function Transition(
+  props: TransitionProps & {
+    children: React.ReactElement;
```
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/extension/src/components/molecules/NFTCard/NFTCard.tsx]: @FlorianBouron it looks like using `ReactNode` here gives typing error, wdyt?

File: packages/extension/src/components/molecules/NFTCard/NFTCard.tsx [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+              effect="blur"
+              src={NFTData?.image}
+              width={150}
+            />
+          ) : (
+            <GemWallet />
+          )}
+          {NFTData ? (
+            <Tooltip title={NFTData.NFTokenID}>
+              <div
```
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/extension/src/components/molecules/NFTCard/NFTCard.tsx]: Done, thank you for the suggestion 👍

File: packages/extension/src/components/molecules/NFTCard/NFTCard.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/extension/src/components/molecules/NFTCard/NFTCard.tsx:29]: So in out `NFTData` structure, we use this standard 👍  But all the properties such as the Attributes, collection, etc are not yet displayed in the UI... We will focus on this once there are NFTs on the market that use the standard

File: packages/extension/src/contexts/LedgerContext/LedgerContext.tsx [hunk 25]
```diff
@@ -90,6 +84,10 @@ interface SubmitTransactionResponse {
   hash: string;
 }
 
+interface NFTImageRequest {
```
[the remark this answers is not in the document]
Comment (ThibautBremand) [reply packages/extension/src/contexts/LedgerContext/LedgerContext.tsx:87]: This makes sense, but this was also because all the methods of the LedgerContext take a typed `payload` object so it keeps things harmonious

File: packages/extension/src/components/molecules/NFTCard/NFTCard.tsx [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+          </Tooltip>
+          <TruncatedText
+            text={NFTData.name}
+            sx={{ fontSize: '16px', color: 'white', marginTop: '10px' }}
+            data-testid="nft_name"
+          />
+          <TruncatedText
+            text={NFTData.description}
+            sx={{ fontSize: '14px', color: 'grey', marginTop: '10px' }}
+            maxLength={200}
```
Comment (florent-uzio) [inline packages/extension/src/components/molecules/NFTCard/NFTCard.tsx:114]: I see what you want to achieve here.
I think it can be improved by letting the CSS magic to handle the display of the multi lines text. Setting the string max length manually is "clunky" in my view.
I would make an update to your current `TruncatedText` component and also create a new component to handle multi lines text truncation.

Here is a code sandbox example with the updated `TruncatedText` and `TruncatedMultiLinesText` component:
https://codesandbox.io/p/sandbox/boring-heisenberg-qrrv9w

The `TruncatedMultiLinesText` component accepts ` maxLines` prop.

In both component you can initialize the `text` prop with `""` to avoid checking the `null` value later. It's cleaner this way.

Let me know what you think.

File: packages/extension/src/components/atoms/TruncatedText/TruncatedText.tsx [hunk 27]
```diff
@@ -0,0 +1,31 @@
+import { ComponentProps, FC } from 'react';
+
+import { Typography } from '@mui/material';
+
+type TruncatedTextProps = ComponentProps<typeof Typography> & {
+  text?: string;
+  maxLength?: number;
+};
+
+export const TruncatedText: FC<TruncatedTextProps> = ({ text, maxLength, sx, ...rest }) => {
```
Comment (florent-uzio) [inline packages/extension/src/components/atoms/TruncatedText/TruncatedText.tsx:10]: See my suggested updates for that component in my sandbox:

https://codesandbox.io/p/sandbox/boring-heisenberg-qrrv9w

File: packages/extension/src/components/atoms/TruncatedText/TruncatedText.tsx [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+  if (!text) return null;
```
Comment (florent-uzio) [inline packages/extension/src/components/atoms/TruncatedText/TruncatedText.tsx:11]: You can initialize the prop with `""` and avoid this `if` block.

**Review by florent-uzio — COMMENTED** [review]
I had a few minutes to add some comments about the TruncatedText component.
Check this sandbox: https://codesandbox.io/p/sandbox/boring-heisenberg-qrrv9w

Let me know what you think.

File: packages/extension/src/components/molecules/NFTCard/NFTCard.tsx (hunk 26 above)
Comment (ThibautBremand) [reply packages/extension/src/components/molecules/NFTCard/NFTCard.tsx]: Thanks a lot, I have refactored the `TruncatedText` component and introduced the `TruncatedMultiLinesText` 👍 It seems great

Comment (ThibautBremand) [reply packages/extension/src/components/molecules/NFTCard/NFTCard.tsx]: Thanks a lot! I have refactored the `TruncatedText` component and introduced the `TruncatedMultiLinesText` 👍.  

I think it's great

File: packages/extension/src/components/atoms/TruncatedMultiLinesText/TruncatedMultiLinesText.tsx [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+type TruncatedMultiLinesTextProps = ComponentProps<typeof Typography> & {
+  text?: string;
+  maxLines?: string;
+};
+
+export const TruncatedMultiLinesText = ({
+  text = '',
+  maxLines = '4',
+  sx,
+  variant = 'body1'
```
Comment (florent-uzio) [inline packages/extension/src/components/atoms/TruncatedMultiLinesText/TruncatedMultiLinesText.tsx:14]: I forgot it yesterday when I sent the sandbox to you, I updated the sandbox now.

But you can replace this "variant" by `...rest` to take all the props from `Typography` and then you pass them to the `Typography` component below.

The full component:

```jsx
import Typography from "@mui/material/Typography";
import { ComponentProps } from "react";

type TruncatedMultiLinesTextProps = ComponentProps<typeof Typography> & {
  text?: string;
  maxLines?: string;
};

export const TruncatedMultiLinesText = ({
  text = "",
  maxLines = "4", // defaults to 4 max lines
  sx,
  ...rest
}: TruncatedMultiLinesTextProps) => {
  return (
    <Typography
      sx={{
        overflow: "hidden",
        display: "-webkit-box",
        "-webkit-box-orient": "vertical",
        "-webkit-line-clamp":
          maxLines /* start showing ellipsis when X line is reached */,
        "white-space": "pre-wrap" /* let the text wrap preserving spaces */,
        ...sx,
      }}
      {...rest}
    >
      {text}
    </Typography>
  );
};
```

Also `variant` is by default `body1`.

File: packages/extension/src/components/atoms/TruncatedText/TruncatedText.tsx [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+    <Typography
+      sx={{
+        display: 'block',
+        overflow: 'hidden',
+        textOverflow: 'ellipsis',
+        width: '100%',
+        ...sx
+      }}
+      {...rest}
+      variant={variant}
```
Comment (florent-uzio) [inline packages/extension/src/components/atoms/TruncatedText/TruncatedText.tsx:25]: No need of this as it's already part of `...rest`, you can remove it.

File: packages/extension/src/components/atoms/TruncatedText/TruncatedText.tsx [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+import { Typography } from '@mui/material';
+
+type TruncatedTextProps = ComponentProps<typeof Typography> & {
+  text?: string;
+};
+
+export const TruncatedText = ({
+  text = '',
+  sx,
+  variant = 'body1',
```
Comment (florent-uzio) [inline packages/extension/src/components/atoms/TruncatedText/TruncatedText.tsx:12]: This can be removed, it's part of rest and by default is `body1`.

File: packages/extension/src/components/organisms/NFTDetails/NFTDetails.tsx [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+            <CloseIcon />
+          </IconButton>
+          <Typography sx={{ ml: 2, flex: 1 }} variant="h6" component="p">
+            NFT Details
+          </Typography>
+        </Toolbar>
+      </AppBar>
+      <List sx={{ width: '100%', wordBreak: 'break-word' }}>
+        <NFTImage imageURL={NFTData.image} />
+        <NFTListItem primary="Token ID" secondary={NFTData.NFTokenID} />
```
Comment (florent-uzio) [inline packages/extension/src/components/organisms/NFTDetails/NFTDetails.tsx:37]: Actually, now that I see it, I am thinking it's maybe not necessary to create that extra `NFTListItem` component... It doesn't simplify much and hardcoding the style directly into `NFTListItem` is not optimized.

We could just extract the `ListItem` style into its own const that we create outside the component and reuse it.

```jsx
const listItemStyle = { padding: "8px 24px" }

export const NFTDetails = () => {
...

return (
...
<ListItem style={listItemStyle}>
     <ListItemText primary={} secondary={} />
</ListItem>
...
```

Sorry for the change :)

File: packages/extension/src/components/atoms/TruncatedMultiLinesText/TruncatedMultiLinesText.tsx (hunk 29 above)
Comment (ThibautBremand) [reply packages/extension/src/components/atoms/TruncatedMultiLinesText/TruncatedMultiLinesText.tsx]: Done, thank you!

File: packages/extension/src/components/organisms/NFTDetails/NFTDetails.tsx (hunk 32 above)
Comment (ThibautBremand) [reply packages/extension/src/components/organisms/NFTDetails/NFTDetails.tsx]: Done 👍

**Review by FlorianBouron — APPROVED** [review]
